askama/src/filters/json.rs

```diff
@@ -19,17 +19,74 @@ use serde_json::to_writer_pretty;
 /// or in apostrophes with the (optional) safe filter `'{{data|json|safe}}'`.
 /// In HTML texts the output of e.g. `<pre>{{data|json|safe}}</pre>` is safe, too.
 #[inline]
-pub fn json<S: Serialize>(s: S) -> Result<impl fmt::Display, Infallible> {
-    Ok(ToJson(s))
+pub fn json<S: Serialize, I: AsIndent>(
+    value: S,
+    indent: I,
+) -> Result<impl fmt::Display, Infallible> {
+    Ok(ToJson { value, indent })
 }
 
-#[derive(Debug, Clone)]
-struct ToJson<S: Serialize>(S);
+pub trait AsIndent {
+    fn as_indent(&self) -> Option<&str>;
+}
+
+impl AsIndent for str {
+    #[inline]
+    fn as_indent(&self) -> Option<&str> {
+        Some(self)
+    }
+}
+
+impl AsIndent for isize {
+    #[inline]
+    fn as_indent(&self) -> Option<&str> {
+        const SPACES: &str = "                ";
+        match *self < 0 {
+            true => None,
+            false => Some(&SPACES[..(*self as usize).min(SPACES.len())]),
+        }
+    }
+}
 
-impl<S: Serialize> fmt::Display for ToJson<S> {
+impl AsIndent for () {
     #[inline]
+    fn as_indent(&self) -> Option<&str> {
+        None
+    }
+}
+
+impl<T: AsIndent + ?Sized> AsIndent for &T {
+    #[inline]
+    fn as_indent(&self) -> Option<&str> {
+        T::as_indent(self)
+    }
+}
+
+impl<T: AsIndent> AsIndent for Option<T> {
+    #[inline]
+    fn as_indent(&self) -> Option<&str> {
+        self.as_ref().and_then(T::as_indent)
+    }
+}
+
+#[derive(Debug, Clone)]
+struct ToJson<S, I> {
+    value: S,
+    indent: I,
+}
+
+impl<S: Serialize, I: AsIndent> fmt::Display for ToJson<S, I> {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        to_writer_pretty(JsonWriter(f), &self.0).map_err(|_| fmt::Error)
+        let f = JsonWriter(f);
+        if let Some(indent) = self.indent.as_indent() {
```

Comment:
Nit: would prefer to `match` this, yielding out the `self.indent` value and doing an early return for the `None` case.

Reply:
No indentation is not the same as no beautification. A Some("") prefix still generates newlines, None does not. So both cases are entirely different. I could replace the `if` with a `match`, though.